calyx-opt/src/analysis/compute_static.rs

```diff
@@ -133,3 +134,90 @@ impl WithStatic for ir::While {
         Some(bound * b_time)
     }
 }
+
+pub trait IntoStatic {
+    type StaticCon;
+    fn make_static(&mut self) -> Option<Self::StaticCon>;
+}
+
+impl IntoStatic for ir::Seq {
+    type StaticCon = ir::StaticSeq;
+    fn make_static(&mut self) -> Option<Self::StaticCon> {
+        let mut static_stmts: Vec<ir::StaticControl> = Vec::new();
+        let mut latency = 0;
+        for stmt in self.stmts.iter() {
+            if !matches!(stmt, ir::Control::Static(_)) {
+                log::debug!("control under seq is not static");
+                return None;
+            }
+        }
+
+        for stmt in self.stmts.drain(..) {
+            let ir::Control::Static(sc) = stmt else {unreachable!("We have already checked that all control statements are static")};
+            latency += sc.get_latency();
+            static_stmts.push(sc);
+        }
+        Some(ir::StaticSeq {
+            stmts: static_stmts,
+            attributes: self.attributes.clone(),
+            latency,
+        })
+    }
+}
+
+impl IntoStatic for ir::Par {
+    type StaticCon = ir::StaticPar;
+    fn make_static(&mut self) -> Option<Self::StaticCon> {
+        let mut static_stmts: Vec<ir::StaticControl> = Vec::new();
+        let mut latency = 0;
+        for stmt in self.stmts.iter() {
+            if !matches!(stmt, ir::Control::Static(_)) {
+                log::debug!("control under par is not static");
+                return None;
+            }
+        }
+
+        for stmt in self.stmts.drain(..) {
+            let ir::Control::Static(sc) = stmt else {unreachable!("We have already checked that all control statements are static")};
+            latency = std::cmp::max(sc.get_latency(), sc.get_latency());
+            static_stmts.push(sc);
+        }
+        Some(ir::StaticPar {
+            stmts: static_stmts,
+            attributes: self.attributes.clone(),
+            latency,
+        })
+    }
+}
+
+impl IntoStatic for ir::If {
+    type StaticCon = ir::StaticIf;
+    fn make_static(&mut self) -> Option<Self::StaticCon> {
+        if let ir::Control::Static(_) = *self.tbranch {
```

Comment:
Use `let`-`else` and add a `return None` statement in the `else` branch to make this code less nested.